File: extensions/reddit/src/main/java/app/morphe/extension/reddit/patches/HideSidebarComponentsPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */
package app.morphe.extension.reddit.patches;

import java.util.Collection;
import java.util.Collections;

import app.morphe.extension.reddit.settings.Settings;
import app.morphe.extension.shared.settings.BooleanSetting;

@SuppressWarnings("unused")
public final class HideSidebarComponentsPatch {
    /**
     * Interface to use obfuscated methods.
     */
    public interface HeaderItemInterface {
        // Methods are added during patching.
        String patch_getItemName();
    }

    /**
     * @return If this patch was included during patching.
     */
    public static boolean isPatchIncluded() {
        return false;  // Modified during patching.
    }

    /**
     * Injection point.
     */
    public static Collection<?> hideComponents(Collection<?> c, HeaderItemInterface headerItemUiModel) {
        if (headerItemUiModel != null && !c.isEmpty()) {
            String headerItemName = headerItemUiModel.patch_getItemName();
            for (HeaderItem headerItem : HeaderItem.values()) {
                if (headerItem.enabled && headerItem.name().equals(headerItemName)) {
                    return Collections.emptyList();
                }
            }
        }

        return c;
    }

    private enum HeaderItem {
        ABOUT(Settings.HIDE_ABOUT_SHELF),
        COMMUNITIES(false),
        COMMUNITY_CLUBS(false),
        COMMUNITY_EVENT(false),
        FAVORITES(false),
        FOLLOWING(false),
        GAMES_ON_REDDIT(Settings.HIDE_GAMES_ON_REDDIT_SHELF),
        MODERATING(false),
        RECENTLY_VISITED(Settings.HIDE_RECENTLY_VISITED_SHELF),
        REDDIT_PRO(Settings.HIDE_REDDIT_PRO_SHELF),
        RESOURCES(Settings.HIDE_RESOURCES_SHELF);

        private final boolean enabled;

        HeaderItem(boolean enabled) {
            this.enabled = enabled;
        }

        HeaderItem(BooleanSetting setting) {
            this.enabled = setting.get();
        }
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/layout/sidebar/HideSidebarComponentsPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */
package app.morphe.patches.reddit.layout.sidebar

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patcher.util.proxy.mutableTypes.MutableMethod.Companion.toMutable
import app.morphe.patches.reddit.misc.settings.settingsPatch
import app.morphe.patches.reddit.shared.Constants.COMPATIBILITY_REDDIT
import app.morphe.util.setExtensionIsPatchIncluded
import com.android.tools.smali.dexlib2.AccessFlags
import com.android.tools.smali.dexlib2.builder.MutableMethodImplementation
import com.android.tools.smali.dexlib2.immutable.ImmutableMethod
import java.util.logging.Logger

internal const val EXTENSION_CLASS =
    "Lapp/morphe/extension/reddit/patches/HideSidebarComponentsPatch;"

private const val EXTENSION_HEADER_ITEM_INTERFACE =
    $$"Lapp/morphe/extension/reddit/patches/HideSidebarComponentsPatch$HeaderItemInterface;"

@Suppress("unused")
val hideSidebarComponentsPatch = bytecodePatch(
    name = "Hide sidebar components",
    description = "Adds options to hide the sidebar components."
) {
    compatibleWith(COMPATIBILITY_REDDIT)

    dependsOn(settingsPatch)

    execute {
        CommunityDrawerBuilderFingerprint.method.apply {
            val collectionParameter = parameterTypes.indexOf("Ljava/util/Collection;")
            addInstructions(
                0,
                """
                    invoke-static/range { p$collectionParameter .. p${collectionParameter + 1} }, $EXTENSION_CLASS->hideComponents(Ljava/util/Collection;$EXTENSION_HEADER_ITEM_INTERFACE)Ljava/util/Collection;
                    move-result-object p$collectionParameter
                """
            )
        }

        HeaderItemUiModelToStringFingerprint.let {
            it.classDef.apply {
                // Add interface and helper methods to allow extension code to call obfuscated methods.
                interfaces.add(EXTENSION_HEADER_ITEM_INTERFACE)
                // Add methods to access obfuscated header item fields.
                methods.add(
                    ImmutableMethod(
                        type,
                        "patch_getItemName",
                        listOf(),
                        "Ljava/lang/String;",
                        AccessFlags.PUBLIC.value or AccessFlags.FINAL.value,
                        null,
                        null,
                        MutableMethodImplementation(2),
                    ).toMutable().apply {
                        val headerItemField = fields.single { field ->
                            field.type == "Lcom/reddit/screens/drawer/community/HeaderItem;"
                        }

                        addInstructionsWithLabels(
                            0,
                            """
                                iget-object v0, p0, $headerItemField
                                if-nez v0, :get_name
                                const-string v0, ""
                                return-object v0
                                :get_name
                                invoke-virtual { v0 }, Ljava/lang/Enum;->name()Ljava/lang/String;
                                move-result-object v0
                                return-object v0
                            """
                        )
                    }
                )
            }
        }

        setExtensionIsPatchIncluded(EXTENSION_CLASS)
    }
}